File: implementations/csharp/protocol/src/Specus.Protocol/PeerEgress/Ipv4Cidr.cs
namespace Specus.Protocol.PeerEgress;

/// <summary>
/// An IPv4 prefix.
/// </summary>
/// <remarks>
/// Parsing is strict on purpose: a prefix whose host bits are set is rejected rather than masked,
/// and a leading-zero octet is rejected rather than tolerated. Runtimes disagree about 010 --
/// decimal ten or octal eight -- and an access rule that means different things in different
/// implementations is worse than one the operator has to rewrite.
/// </remarks>
public readonly record struct Ipv4Cidr(uint Network, int PrefixLength)
{
    /// <summary>The widest IPv4 prefix length.</summary>
    public const int MaxPrefix = 32;

    /// <summary>
    /// Reads an IPv4 address or prefix. A bare address becomes a /32.
    /// </summary>
    public static bool TryParse(string? text, out Ipv4Cidr cidr)
    {
        cidr = default;
        var trimmed = text?.Trim() ?? string.Empty;
        if (trimmed.Length == 0)
        {
            return false;
        }
        var slash = trimmed.IndexOf('/');
        var addressPart = slash < 0 ? trimmed : trimmed[..slash];
        if (!TryParseAddress(addressPart, out var address))
        {
            return false;
        }
        if (slash < 0)
        {
            cidr = new Ipv4Cidr(address, MaxPrefix);
            return true;
        }
        var prefixPart = trimmed[(slash + 1)..];
        if (prefixPart.Length is 0 or > 2)
        {
            return false;
        }
        if (prefixPart.Length > 1 && prefixPart[0] == '0')
        {
            return false;
        }
        var prefix = 0;
        foreach (var c in prefixPart)
        {
            if (c is < '0' or > '9')
            {
                return false;
            }
            prefix = (prefix * 10) + (c - '0');
        }
        if (prefix > MaxPrefix)
        {
            return false;
        }
        if ((address & ~MaskFor(prefix)) != 0)
        {
            return false;
        }
        cidr = new Ipv4Cidr(address, prefix);
        return true;
    }

    /// <summary>
    /// Reads four canonical dotted-decimal octets.
    /// </summary>
    public static bool TryParseAddress(string? text, out uint address)
    {
        address = 0;
        if (text is null)
        {
            return false;
        }
        uint value = 0;
        var octets = 0;
        var start = 0;
        for (var i = 0; i <= text.Length; i++)
        {
            if (i != text.Length && text[i] != '.')
            {
                continue;
            }
            var digits = i - start;
            if (digits is < 1 or > 3 || octets > 3)
            {
                return false;
            }
            if (digits > 1 && text[start] == '0')
            {
                return false;
            }
            var part = 0;
            for (var j = start; j < i; j++)
            {
                var c = text[j];
                if (c is < '0' or > '9')
                {
                    return false;
                }
                part = (part * 10) + (c - '0');
            }
            if (part > 255)
            {
                return false;
            }
            value = (value << 8) | (uint)part;
            octets++;
            start = i + 1;
        }
        if (octets != 4)
        {
            return false;
        }
        address = value;
        return true;
    }

    /// <summary>Renders an address in dotted-decimal form.</summary>
    public static string FormatAddress(uint value) =>
        $"{(value >> 24) & 0xFF}.{(value >> 16) & 0xFF}.{(value >> 8) & 0xFF}.{value & 0xFF}";

    /// <summary>Reports whether the address falls inside the prefix.</summary>
    public bool Contains(uint address)
    {
        var mask = MaskFor(PrefixLength);
        return (address & mask) == (Network & mask);
    }

    /// <summary>Reports whether the two prefixes share any address.</summary>
    public bool Overlaps(Ipv4Cidr other)
    {
        var mask = MaskFor(Math.Min(PrefixLength, other.PrefixLength));
        return (Network & mask) == (other.Network & mask);
    }

    /// <summary>Renders the prefix in canonical form.</summary>
    public override string ToString() => $"{FormatAddress(Network)}/{PrefixLength}";

    internal static bool ContainedIn(uint address, IEnumerable<string> cidrs)
    {
        foreach (var text in cidrs)
        {
            if (TryParse(text, out var cidr) && cidr.Contains(address))
            {
                return true;
            }
        }
        return false;
    }

    private static uint MaskFor(int prefixLength) =>
        prefixLength == 0 ? 0u : uint.MaxValue << (MaxPrefix - prefixLength);
}

File: implementations/java/common/src/main/java/com/theshuai/common/peeregress/Ipv4Cidr.java
package com.theshuai.common.peeregress;

/**
 * IPv4 prefix used by both consumer rule matching and egress authorization.
 *
 * <p>Parsing is deliberately strict: a prefix whose host bits are set is rejected rather than
 * masked. Implementations normalise differently, and a rule that silently changes meaning between
 * runtimes is worse than one the user has to rewrite.
 */
public record Ipv4Cidr(int network, int prefixLength) {
    public static final int MAX_PREFIX = 32;

    /** Returns null when the text is not a canonical IPv4 address or prefix. */
    public static Ipv4Cidr parse(String text) {
        if (text == null) {
            return null;
        }
        String trimmed = text.trim();
        if (trimmed.isEmpty()) {
            return null;
        }
        int slash = trimmed.indexOf('/');
        String addressPart = slash < 0 ? trimmed : trimmed.substring(0, slash);
        Integer address = parseAddress(addressPart);
        if (address == null) {
            return null;
        }
        if (slash < 0) {
            return new Ipv4Cidr(address, MAX_PREFIX);
        }
        String prefixPart = trimmed.substring(slash + 1);
        if (prefixPart.isEmpty() || prefixPart.length() > 2) {
            return null;
        }
        int prefix = 0;
        for (int i = 0; i < prefixPart.length(); i++) {
            char c = prefixPart.charAt(i);
            if (c < '0' || c > '9') {
                return null;
            }
            prefix = prefix * 10 + (c - '0');
        }
        if (prefixPart.length() > 1 && prefixPart.charAt(0) == '0') {
            return null;
        }
        if (prefix > MAX_PREFIX) {
            return null;
        }
        if ((address & ~maskFor(prefix)) != 0) {
            // Host bits set. Rejected rather than masked; see the class comment.
            return null;
        }
        return new Ipv4Cidr(address, prefix);
    }

    /**
     * Returns null when the text is not four canonical dotted-decimal octets.
     *
     * <p>A leading zero is rejected rather than tolerated. Runtimes disagree about {@code 010}:
     * some read decimal ten, some read octal eight, and an address that changes meaning between
     * implementations is exactly the kind of ambiguity an access rule must not carry.
     */
    public static Integer parseAddress(String text) {
        if (text == null) {
            return null;
        }
        int value = 0;
        int octets = 0;
        int start = 0;
        int length = text.length();
        for (int i = 0; i <= length; i++) {
            if (i != length && text.charAt(i) != '.') {
                continue;
            }
            int digits = i - start;
            if (digits < 1 || digits > 3 || octets > 3) {
                return null;
            }
            if (digits > 1 && text.charAt(start) == '0') {
                return null;
            }
            int part = 0;
            for (int j = start; j < i; j++) {
                char c = text.charAt(j);
                if (c < '0' || c > '9') {
                    return null;
                }
                part = part * 10 + (c - '0');
            }
            if (part > 255) {
                return null;
            }
            value = (value << 8) | part;
            octets++;
            start = i + 1;
        }
        return octets == 4 ? value : null;
    }

    public static String format(int address) {
        return ((address >>> 24) & 0xFF) + "." + ((address >>> 16) & 0xFF) + "."
                + ((address >>> 8) & 0xFF) + "." + (address & 0xFF);
    }

    private static int maskFor(int prefixLength) {
        return prefixLength == 0 ? 0 : (int) (0xFFFFFFFFL << (MAX_PREFIX - prefixLength));
    }

    public boolean contains(int address) {
        int mask = maskFor(prefixLength);
        return (address & mask) == (network & mask);
    }

    public boolean overlaps(Ipv4Cidr other) {
        if (other == null) {
            return false;
        }
        int shared = Math.min(prefixLength, other.prefixLength);
        int mask = maskFor(shared);
        return (network & mask) == (other.network & mask);
    }

    @Override
    public String toString() {
        return format(network) + "/" + prefixLength;
    }
}
